### PR TITLE
Cleanup C++ model example and cmake CI job

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR cleans up some CMake stuff. I added comments to explain the changes made to the respective files.

This does not make any change to the `CMakeLists.txt` at the root of the repo, and doesn't change the way `libtorchvision` is built either. Those are only changes related to docs on how to build it, and related to the examples / tests we have.

The main changes are:

- remove stuff from `test/tracing/` and integrate them within `example/cpp`
- remove unnecessary dependency on libpython for the examples
- properly depend on libtorchvision in the example but *only* when needed - some models require it, some don't. Now we let the user choose whether they want it or not. This is better documented as well.

### CODE REVIEW

File: .github/scripts/cmake.sh
Comment (NicolasHug) [inline .github/scripts/cmake.sh:1]: This is executed by our Cmake CI job.

Originally this does the following:

- build `torchvisionlib`
- build and run the `example/cpp` stuff
- build and run the `test/tracing/frcnn` stuff.


In this PR I  removed the `test/tracing/frcnn` stuff and merged it into the `example/cpp` stuff, so in this file we're just removing the things related to  `test/tracing/frcnn`.

File: .github/scripts/cmake.sh [hunk 1]
```diff
@@ -85,40 +80,24 @@ fi
...
-
 echo '::group::Build and run C++ example'
-pushd examples/cpp/hello_world/build
+pushd examples/cpp/build
 
 cmake .. -DTorch_DIR="${Torch_DIR}" \
   -DCMAKE_PREFIX_PATH="${CONDA_PREFIX}" \
-  -DCMAKE_FIND_FRAMEWORK=NEVER
+  -DCMAKE_FIND_FRAMEWORK=NEVER \
+  -DUSE_TORCHVISION=ON  # Needed for faster-rcnn since it's using torchvision ops like NMS.
```
Comment (NicolasHug) [inline .github/scripts/cmake.sh:89]: Some scripted models like Faster-RCNN rely on torchvision ops, so for those we need to link to `libtorchvision`. But some other models like resnet do not need `libtorchvision`.

Prior to this PR, the example's `CmakeLists.txt` would  always link to `libtorchvision` even if it was unnecessary. I added a `USE_TORCHVISION` option to let use decide whether they want to link against it or not. Here, we want to test the RCNN model, so we have to pass `USE_TORCHVISION=ON`.

File: examples/cpp/CMakeLists.txt
Comment (NicolasHug) [inline examples/cpp/CMakeLists.txt:1]: This file is a merge of the previous `hello_world/CMakeLists.txt` (now deleted) and `test/tracing/frcnn/CMakelists.txt` (now deleted).

The main changes are:

- only link to libtorchvision if `USE_TORCHVISION` is ON.
- Do not link to libPython - this is actually not needed in this example regardless of the platform (it's only potentially required on `libtorchvision` on Windows).

File: examples/cpp/README.rst
Comment (NicolasHug) [inline examples/cpp/README.rst:1]: Much more detailed and more accurate build instructions for libtorchvision and for the example.

File: examples/cpp/run_model.cpp
Comment (NicolasHug) [inline examples/cpp/run_model.cpp:1]: This is a merge of the previous `hello-world/main.cpp` (deleted) and `test/tracing/frcnn/test_frcnn_tracing.cpp` (deleted).

This file is able to run both a resnet and faster-rcnn scripted model, and only include relevant headers if needed.

File: examples/cpp/run_model.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#include <torch/script.h>
+#include <torch/torch.h>
+#include <cstring>
+#include <iostream>
+
+#ifdef _WIN32
+#include <torchvision/vision.h>
+#endif // _WIN32
```
Comment (NicolasHug) [inline examples/cpp/run_model.cpp:8]: I don't fully understand yet why this is required on Windows, but this is for another PR

File: examples/cpp/script_model.py
Comment (NicolasHug) [inline examples/cpp/script_model.py:1]: like other files this is a merged of the previous stuff that was in `hello-world` and in `tests/tracing`.

**Review by vfdev-5 — APPROVED** [review]
Stamp

**Pull request merged**